xapian-applications/omega/handler_gmime.cc

```diff
@@ -0,0 +1,183 @@
+{
+    GMimeContentType* ct = g_mime_object_get_content_type(me);
+    if (GMIME_IS_MULTIPART(me)) {
+	GMimeMultipart* mpart = (GMimeMultipart*)me;
+	string subtype = g_mime_content_type_get_media_subtype(ct);
+	int count = g_mime_multipart_get_count(mpart);
+	if (subtype == "alternative") {
+	    for (int i = 0; i < count; ++i) {
+		GMimeObject* part = g_mime_multipart_get_part(mpart, i);
+		if (parser_content(part, dump))
+		    return true;
+	    }
+	} else {
+	    bool ret = false;
+	    for (int i = 0; i < count; ++i) {
+		GMimeObject* part = g_mime_multipart_get_part(mpart, i);
+		ret |= parser_content(part, dump);
+	    }
+	    return ret;
+	}
+    } else if (GMIME_IS_PART(me)) {
+	GMimePart* part = (GMimePart*)me;
+	GMimeDataWrapper* content = g_mime_part_get_content_object(part);
+	string type = g_mime_content_type_get_media_type(ct);
+	string subtype = g_mime_content_type_get_media_subtype(ct);
+	string charset(g_mime_object_get_headers(me));
+	size_t end, start;
+	start = charset.find("charset=");
+	if (start != string::npos) {
+	    start += 8;
+	    if (charset[start] == '"')
+		start++;
+	    for (end = start; end < charset.length() &&
+		 !isspace(charset[end]) && charset[end] != '"'; ++end);
+	    charset = charset.substr(start, end - start);
+	    charset = g_mime_charset_canon_name(charset.c_str());
+	} else {
+	    charset.clear();
+	}
+	if (type == "text") {
+	    char buffer[SIZE];
+	    GMimeStream* sr = g_mime_data_wrapper_get_stream(content);
+	    size_t len = g_mime_stream_read(sr, buffer, SIZE);
+	    auto encoding = g_mime_part_get_content_encoding(part);
+	    len = decode((unsigned char*)buffer, len, encoding);
+	    if (subtype == "plain") {
+		if (!charset.empty() && charset != "UTF-8") {
+		    GMimeFilter* ch = g_mime_filter_charset_new(charset.c_str(),
+								"UTF-8");
+		    if (GMIME_IS_FILTER(ch)) {
+			size_t presize = 0;
+			g_mime_filter_complete(ch, buffer, len, presize,
+					      (char**)&buffer, &len, &presize);
+		    }
+		}
+		dump.append(buffer, len);
+	    } else if (subtype == "html") {
+		extract_html((const char*)buffer, charset, dump);
+	    }
+	    return true;
+	}
+    }
+    return false;
+}
+
+bool
+extract(const string& filename,
+	string& dump,
+	string& title,
+	string& keywords,
+	string& author,
+	string& pages,
+	string& error)
+{
+    static bool first_time = true;
+    if (first_time) {
+	g_mime_init(0);
+	first_time = false;
+    }
+
+    FILE* fp = fopen(filename.c_str(), "r");
+
+    if (fp == NULL) {
+	error = "Gmime Error: fail open " + filename;
+	return false;
+    }
+
+    GMimeStream* stream = g_mime_stream_file_new(fp);
+    GMimeParser* parser = g_mime_parser_new_with_stream(stream);
+    GMimeMessage* message = g_mime_parser_construct_message(parser);
+    GMimeObject* body = g_mime_message_get_body(message);
+    author = g_mime_message_get_sender(message);
+    title = g_mime_message_get_subject(message);
+
+    parser_content(body, dump);
+    (void)pages;
```

Comment:
I forget `fclose(fp)`

Reply:
You shouldn't call that - see:

https://developer.gnome.org/gmime/stable/GMimeStreamFile.html#g-mime-stream-file-new

> Note: The created GMimeStreamFile object will own the FILE pointer passed in.

So that should close it for you - if you also close it, that's undefined behaviour - from `man fclose`:

> The  behaviour  of  fclose() is undefined if the stream parameter is an illegal pointer, or is a descriptor already passed to a previous invocation of fclose().